### PR TITLE
Encrypt network communication in JGroups

### LINKED ISSUE #25702: Encrypt network communication in JGroups
### Description

The JGroups communication is not secure between Keycloak servers. JGroups support SSL sockets (http://jgroups.org/manual5/index.html#TLS) but it requires a customized Infinispan XML configuration file.

The work to be done under this issue is to add options to configure the required attributes (basically the keystore and the truststore) and, programmatically, inject the SSL sockets factories in JGroups independently from the Infinispan XML. This way, it works with the default configuration, customs ones, and with the Keycloak Operator in OpenShift. 

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by pruivo**
Fixes #25702

Draft PR because I would like to receive feedback from the following
* Is it okay for these options to live in the `cache` category?
* Where to document how to configure it with the Keycloak Operator?

### CODE REVIEW

**Comment by pruivo** [comment]
I tested with the current Keycloak Operator code locally using `kind` and with the following snipped
```yaml
spec:
...
  additionalOptions:
    - name: jgroups-tls-enabled
      value: "true"
    - name: jgroups-tls-protocol
      value: "TLSv1.3"
    - name: jgroups-tls-keystore-file
      value: /etc/jgroups-tls/keystore.p12
    - name: jgroups-tls-keystore-password
      secret:
        name: keycloak-jgroups-tls-secret
        key: keystorePassword
    - name: jgroups-tls-keystore-type
      value: "pkcs12"
    - name: jgroups-tls-keystore-alias
      value: "site1"
    - name: jgroups-tls-truststore-file
      value: /etc/jgroups-tls/truststore.p12
    - name: jgroups-tls-truststore-password
      secret:
        name: keycloak-jgroups-tls-secret
        key: truststorePassword
    - name: jgroups-tls-keystore-type
      value: "pkcs12"
    - name: jgroups-tls-client-auth
      value: "need"
  instances: 2  
  unsupported:
    podTemplate:
      spec:
        containers:
          - volumeMounts:
              - name: jgroups-tls-volume
                mountPath: /etc/jgroups-tls/
                readOnly: true
        volumes:
          - name: jgroups-tls-volume
            secret:
              secretName: keycloak-jgroups-tls-secret
```
The secret was created using 
```
kubectl create secret generic keycloak-jgroups-tls-secret \
    --from-file=keystore.p12="./site1-keystore.p12" \
    --from-literal=keystorePassword=*** \
    --from-file=truststore.p12="./ca-keystore.p12" \
    --from-literal=truststorePassword=*** \
```

It is probably worth to write these example in some documentation page :+1:

File: quarkus/config-api/src/main/java/org/keycloak/config/CachingOptions.java
Comment (ahus1) [inline quarkus/config-api/src/main/java/org/keycloak/config/CachingOptions.java:72]: Do you know that people actually do / need to do this in their environment? If it is not a common thing to do, I'd suggest removing this parameter.

File: quarkus/config-api/src/main/java/org/keycloak/config/CachingOptions.java
Comment (ahus1) [inline quarkus/config-api/src/main/java/org/keycloak/config/CachingOptions.java:125]: From what we discussed earlier we want to prevent unauthorized clients to join the communication. So the only option that makes sense IMHO is `need` here. If anything else is not a common use case, we should remove this option and hard-code it to `need`. A user of Keycloak can still change the XML if they want to do something similar. 

If we allow for this option to be changed, I assume people will choose an unsecure setting by accident.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (ahus1) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:148]: Would it be correct to assume that TLS works only with TCP as a transport portocol? If yes, please add a check for that, and throw an exception if a mismatch is detected.

**Review by ahus1 — CHANGES_REQUESTED** [review]
Thank you for this PR - I'm looking forward to see the communication finally encrypted :-)

The documentation from a CLI perspective should probably go here https://www.keycloak.org/server/caching#_securing_cache_communication. Adding it to the list of things to do in https://www.keycloak.org/server/configuration-production would also be good.

I'm not sure we then need additional documentation for the Operator. Once we've settled on the docs and the options, I'd like to see them in the HA guide eventually.

I see the options are currently named `jgroups-tls-...`. As users of Keycloak know on "the cache", I'd suggest to rename it `cache-tls-...` and hide JGroups as an implementation detail.

I still want to test this manually in a cluster, and we would like to see a test case for this in Keycloak before we can merge this.

**Comment by ahus1** [comment]
Hi @keycloak/cloud-native team - with @pruivo knowledge about Infinispan, we would like offer Keycloak CLI options to enable encrypted JGroup communication used for Infinispan between nodes. 

In a nutshell: create a self-signed certificate, add it to the CLI to use it for all internal JGroups communications of the embedded cache, and be safe.

A thing to watch out for: It need be a self-signed certificate used only for this purpose, as only those who have a client certificate signed with that CA would should be allowed to connect. The good thing: no other permissions need to configured for this. It must not use a CA used also for other purposes, as that would allow a lot of other clients to connect as well. 

We'd like to get your early feedback on this PR, for example on the naming of the CLI options. We know that docs and a test are still missing. 

I'm not sure if we would like to see a test Keycloak-only, or one that is using the Operator.

**Comment by ahus1** [comment]
Update: Manual test where I deployed this on our OpenShift cluster worked nicely.

File: quarkus/config-api/src/main/java/org/keycloak/config/CachingOptions.java
Comment (pruivo) [reply quarkus/config-api/src/main/java/org/keycloak/config/CachingOptions.java:72]: We don't allow custom providers in Infinispan Server but we have the option to use the JDK provider or the OpenSSL provider (via WildFly-OpenSSL). I'm going to remove this option.

File: quarkus/config-api/src/main/java/org/keycloak/config/CachingOptions.java
Comment (pruivo) [reply quarkus/config-api/src/main/java/org/keycloak/config/CachingOptions.java:125]: Makes sense.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (pruivo) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:148]: Yes, this only works with TCP sockets. UDP sockets or Java NIO channels are not supported.

**Comment by pruivo** [comment]
Updated
* Added some docs that require some love
* Renamed `jgroups-tls` with `cache-tls`
* Added validation to check if the protocol used supports TLS or not

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java
Comment (ahus1) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java]: I saw your latest changes for this. I suppose it works, still I wonder why this is doing a several negative checks (things that shouldn't be present) instead of a positive check that it is using TCP.

**Comment by ahus1** [comment]
Agreed namings with @vmuzikar in today's call + reduced list to avoid too many configuration options: 

```
    - name: cache-embedded-mtls-enabled
      value: "true"
    - name: cache-embedded-mtls-key-store-file
      value: /etc/jgroups-tls/keystore.p12
    - name: cache-embedded-mtls-key-store-password
      secret:
        name: keycloak-jgroups-tls-secret
        key: keystorePassword
    - name: cache-embedded-mtls-key-store-alias
      # HTTP server works without a key alias, can we get rid of the option? 
      value: "site1"
    - name: cache-embedded-mtls-trust-store-file
      value: /etc/jgroups-tls/truststore.p12
    - name: cache-embedded-mtls-trust-store-password
      secret:
        name: keycloak-cache-tls-secret
        key: truststorePassword
```

In the future, the Operator could possibly provide the truststore and keystore without the need for the user to provide them.

**Comment by sschu** [comment]
Ideally using https://cert-manager.io/.

**Comment by ahus1** [comment]
@sschu - IMHO this doesn't work here, as we need a private CA per KC instance for this. Using a public CA would allow all certificates issued by that CA access to the Keycloak Infinispan setup, as the setup would accept any client certificate signed by the CA.

**Comment by ahus1** [comment]
@pruivo - switching the internal communication to TLS, would it be affected by the bugs discovered around https://github.com/keycloak/keycloak-benchmark/issues/675? 

That is: Once we switch the internal JGroups communication to TLS, it might stall in a similar way if a Keycloak server is not responding, or would it behave better due to the failure detection mechanism?

**Comment by pruivo** [comment]
@ahus1 I would have to test it. Will do that next Monday since I have a doctor appointment this afternoon.

**Comment by ahus1** [comment]
Sure, no hurry. Just a thought and I'm curious. Let's solve the Gossip router situation first.

**Comment by sschu** [comment]
@ahus1 You can use multiple CA's with cert-manager, if I am not mistaken: https://cert-manager.io/docs/configuration/ca/

**Comment by ahus1** [comment]
@pruivo - I've updated the parameters to match the suggestions by @vmuzikar, by adding additional `-`, and removed those were he would to use the default values always.

One parameter remains that should be removed: Can we get rid of the keystore alias? Would it "just work" without? In all other examples (for example the key for HTTPS) we don't pass it, so we would like to get rid of that parameter as well.

**Comment by pruivo** [comment]
@ahus1 I removed the alias. Tested with a self-signed keystore that I use for testing and with a keystore generated from cert-manager.

**Comment by ahus1** [comment]
@pruivo - I reviewed the PR and cleaned up the properties. 

When testing it, I found it would be helpful to add a log message to say that MTLS is now enabled, as mis-typing the parameters wouldn't throw any error when using the operator. I also found that the 4 parameters for keystore/truststore should be required, again for avoiding mistyped parameters. I found out the hard way that it would use the default keystore/truststore in that case, and that would lead most likely to a misconfiguration as this is supposed to be used with a self-signed certificate in our case. 

I also rebased it and updated the help files to the latest state.

I tested this with PR https://github.com/keycloak/keycloak-benchmark/pull/685 on KCB and OpenShift with a locally built Keycloak image.

As I did some changes, I'd like you to have a look at my commit. If you agree, please comment here, or add another commit with any changes you'd like to see.

As the CLI options now align with what we agreed with @vmuzikar, I assume it is fine to be merged afterwards. 

Thanks!

**Review by ahus1 — APPROVED** [review]
Worked in my tests after the review changes. Waiting for review of my changes by @pruivo

**Comment by pruivo** [comment]
@ahus1 is looks good to me :+1:

**Comment by vmuzikar** [comment]
@pruivo @ahus1 Thanks for the changes! I plan to review tomorrow, or early next week.

**Pull request merged**
